[clang-tidy] Add readability-use-span-first-last check

hjanuschka opened the pull request:
Add new clang-tidy check that suggests using std::span's more expressive first() and last() member functions instead of equivalent subspan() calls.


These dedicated methods were added to the standard to provide clearer alternatives to common subspan operations. They improve readability by better expressing intent and are less error-prone by eliminating manual offset calculations.

For example:
```cpp
std::span<int> s = ...;
auto sub1 = s.subspan(0, n);           // transforms to: auto sub1 = s.first(n);
auto sub2 = s.subspan(s.size() - n);   // transforms to: auto sub2 = s.last(n);
auto sub3 = s.subspan(1, n);           // not transformed, no clearer alternative
auto sub4 = s.subspan(n);              // not transformed, no clearer alternative
```


not sure if this is a readability or a modernize check ❓

carlosgalvezp commented:
> not sure if this is a readability or a modernize check 

`readability`. `modernize` is for using new features from newer standards.

carlosgalvezp commented:
```cpp
auto sub2 = s.subspan(n);        // transforms to: auto sub2 = s.last(s.size() - n);
```

IMHO the transformed code is less readable. It would be more useful to do the opposite transformation, maybe that's what you intended?

```cpp
auto sub2 = s.subspan(s.size() - n);        // transforms to: auto sub2 = s.last(n);
```

carlosgalvezp commented on clang-tools-extra/clang-tidy/modernize/UseSpanFirstLastCheck.h:
```diff
@@ -0,0 +1,40 @@
+/// \endcode
+class UseSpanFirstLastCheck : public ClangTidyCheck {
+public:
+  UseSpanFirstLastCheck(StringRef Name, ClangTidyContext *Context)
```
Limit this check to C++20

hjanuschka commented:
@carlosgalvezp thanks for the feedback. I've removed the transformation:

```c++
auto sub2 = s.subspan(n);        // transforms to: auto sub2 = s.last(s.size() - n);
```

While I agree that the transformed expression is more complex (due to size() - x), I initially included it for two reasons:

Consistency: Using `first()/last()` everywhere possible makes the codebase more uniform
The method name `last()` better expresses the intent than `subspan()`, even if the calculation is more verbose

Let me know if you think we should revisit this decision.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,23 @@
+
+Covered scenarios:
+
+==================================== ==================================
```
Columns width.

hjanuschka replied:
Done.

EugeneZelenko commented on clang-tools-extra/test/clang-tidy/checkers/readability/use-span-first-last.cpp:
```diff
@@ -0,0 +1,51 @@
+
+  auto sub4 = s.subspan(1, 2);  // No warning
+  auto sub5 = s.subspan(2);     // No warning
+}
```
Missing newline.

hjanuschka replied:
Done.

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,110 @@
+
+  if (!Replacement.empty()) {
+    if (IsZeroOffset && Count) {
+      diag(Call->getBeginLoc(), "prefer span::first() over subspan()")
+          << FixItHint::CreateReplacement(Call->getSourceRange(), Replacement);
+    } else {
+      diag(Call->getBeginLoc(), "prefer span::last() over subspan()")
+          << FixItHint::CreateReplacement(Call->getSourceRange(), Replacement);
```
Names from code should be put into `'` inside of diagnostics. (Because it's the default behavior when streaming named AST nodes into diagnostics)

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-starts-ends-with.rst:
You removed this by mistake, it's from another check

hjanuschka replied:
🙈

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace clang::tidy::readability {
+
+/// Suggests using clearer std::span member functions first()/last() instead of
+/// equivalent subspan() calls where applicable.
```
Please escape the function/class names with `'`

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  // Build replacement text
+  std::string Replacement;
```
This replacement should be possible to do without the need to go through the lexer, and just replace the needed parts instead. As in, replace from `substr` until before the `n` with `first(/last(` (though that would be problematic with a macro that does `#define K s.size() - 3`, because the `3` would be inside the macro (location wise))...

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/readability/use-span-first-last.cpp:
Please add tests with macros involved (e.g., fully wrapped `#define F(obj, n)...` or `#define N 2`, etc), and with templates

5chmidti reviewed (changes requested):
It would be great to have this check work with uninstantiated templates and use `TK_TraversUnlessSpelledInSource` as well.

Here is an example of a matcher for the uninstantiated case: 
https://godbolt.org/z/1xnGGj7z9

I couldn't fully write it, because there is no matcher to get the member information and constrain it to `subspan`.

---

Please also adjust your PR comment to reflect the change for what pattern is matched in the `last` case.

---

Overall a good readability improvement check, nice

hjanuschka commented on clang-tools-extra/clang-tidy/readability/CMakeLists.txt:
```diff
@@ -58,6 +58,7 @@ add_clang_library(clangTidyReadabilityModule STATIC
   UppercaseLiteralSuffixCheck.cpp
   UseAnyOfAllOfCheck.cpp
   UseStdMinMaxCheck.cpp
+  UseSpanFirstLastCheck.cpp
```
fixed, with the 2 "Use" (https://github.com/llvm/llvm-project/pull/118074/files/c7fd4d5f508d43cd7fdaf7ed9a0006687a4215ed#diff-4be36b5009d110fd7bf319b8b06cf0e86d91b04f4c30dcd8aaf94825626b2e2aL26)  filenames, but there is `RedundantInlineSpecifierCheck`  also wrong, should i fix this in the same PR?

EugeneZelenko replied:
I think one-line change is fine in this pull request.

hjanuschka reviewed (commented):
MANY thanks, yet again, feedback addressed.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+readability-use-span-first-last
+===============================
+
+Checks for uses of ``std::span::subspan()`` that can be replaced with clearer
```
Please synchronize first statement with Release Notes.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+
+Covered scenarios:
+
+========================== ====================
```
Please enlarge first column to `s.subspan(s.size() - n)` (with all back-ticks) width. Second column is too wide.

hjanuschka replied:
rst formatting is really tough 😬

hjanuschka replied:
Done.

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+  const Expr *Count = NumArgs > 1 ? Call->getArg(1) : nullptr;
+  auto &Context = *Result.Context;
+
+  class SubspanVisitor : public RecursiveASTVisitor<SubspanVisitor> {
```
(on mobile, so I can't check)

The visitor is overkill and will, most importantly, produce false positives: `s.subspan(0, 100 + (s.size() - n))`, because the visitor only searches for the specific pattern being contained in the arg, and not that it is actually the arg itself. It would be much cleaner to check this in the matcher: `anyOf(callExpr(has argument(0, integer literal(equals(0)).bind("zero_offset"), /*n arg*/), callExpr(/*size-n arg*/)`

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+        CharSourceRange::getTokenRange(Count->getSourceRange()),
+        Context.getSourceManager(), Context.getLangOpts());
+    const auto *Base =
+        cast<CXXMemberCallExpr>(Call)->getImplicitObjectArgument();
```
`Call` is already a `CXXMemberCallExpr`

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+        CharSourceRange::getTokenRange(Visitor.SizeMinusArg->getSourceRange()),
+        Context.getSourceManager(), Context.getLangOpts());
+    const auto *Base =
+        cast<CXXMemberCallExpr>(Call)->getImplicitObjectArgument();
```
Same as above

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+(like ``subspan(n)``) have no clearer equivalent using ``first()`` or 
+``last()``, so these cases are not transformed.
+
+This check is only active when C++20 or later is used.
```
Missing newline

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/readability/use-span-first-last.cpp:
```diff
@@ -0,0 +1,98 @@
+// Test instantiation
+void testInt() {
+  testTemplate<int>();
+}
```
Missing newline

hjanuschka replied:
Done.

hjanuschka reviewed (commented):
@5chmidti removed the recursive visitor and refactored it, hope this is correct now, added the test cases, you mentioned, and you where right in the complex version they failed, work now!

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+
+Covered scenarios:
+
+=============================== ============
```
Please reduce first column on 4 characters and add two characters to second column, so all cell content will be covered.

hjanuschka replied:
resolved.

hjanuschka commented:
@EugeneZelenko  CI fails because of 
> ReleaseNotes.rst:145:unknown document: 'clang-tidy/checks/readability/readability-use-span-first-last'

could you help me, i don't get it why this fails, bet its  a small thing but i am too blind for this. 

also, since i have troubles doing the RST's  is there any documentation on how to run the "docs-CI" locally (to prevent me from pushing wrong commits back and forth 🙈 in the hope to get a green light)

carlosgalvezp commented:
You have duplicated `readability` when referring to the document:

```
- 'clang-tidy/checks/readability/readability-use-span-first-last'
+ 'clang-tidy/checks/readability/use-span-first-last'
```

You can build them with: `ninja docs-clang-tools-html`, with following CMake config:

```
cmake -B build -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" -DLLVM_ENABLE_SPHINX=ON ./llvm
```

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -18,108 +18,77 @@ using namespace clang::ast_matchers;
+          .bind("first_subspan"),
+      this);
+
+  // Match span.subspan(size() - n) -> last(n)
```
nit: `span.size()`

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -18,108 +18,77 @@ using namespace clang::ast_matchers;
+      this);
+
+  // Match span.subspan(size() - n) -> last(n)
+  const auto SizeCall = cxxMemberCallExpr(
+      callee(memberExpr(hasDeclaration(cxxMethodDecl(hasName("size"))))));
```
The size could also be taken via `std::[ranges::]size`, or if the span is `const`, then also from a `const` variable:

```c++
void foo(const std::span<int> sp) {
  const auto s = std::ranges::size(sp);
  sp.subspan(s - 10);
}
```

What do others think?

hjanuschka replied:
thanks, great feedback, added matchers and test cases for ranges

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -18,108 +18,77 @@ using namespace clang::ast_matchers;
+        *Result.SourceManager, Result.Context->getLangOpts());
+
+    std::string Replacement =
+        SpanText.str() + ".first(" + CountText.str() + ")";
```
Please use a twine instead of adding `std::string`s together: `(SpanTExt + ".first(" + CountText + ")").str()`

hjanuschka replied:
Done.

HerrCai0907 commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,94 @@
+  if (const auto *LastCall =
+          Result.Nodes.getNodeAs<CXXMemberCallExpr>("last_subspan")) {
+    const auto *Count = Result.Nodes.getNodeAs<Expr>("count");
+    if (!Count)
+      return;
```
can be assert

HerrCai0907 reviewed (commented):
It may still work on `TraversalKind::TK_IgnoreUnlessSpelledInSource`. Could you try to set traversal mode in `getCheckTraversalKind`. It can avoid noise of something like implicit cast.

HerrCai0907 commented:
> tried using `TK_IgnoreUnlessSpelledInSource` but this breaks the template based cases.

Thanks! If it will waste you more effort, it's good to keep it like this.

hjanuschka commented:
wasting effort would not be the problem (it never is; learning is the goal) - right now i just dont understand the difference, and therefore dont see the benefit, as far as i see it would introduce more complicated code to still handle templates.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+
+Covered scenarios:
+
+=========================== ============
```
Second column should be expanded on two characters to fit `s.first(n)` with all back-ticks.

hjanuschka replied:
thank you, super sorry that this is such a back n forth, but my brain seems to be wired a way to not being able to figure RST out 🙈

EugeneZelenko replied:
You just need to look on diff :-)

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+  // Match span.subspan(size() - n) or span.subspan(std::ranges::size(span) - n)
+  // -> last(n)
+  const auto SizeCall = anyOf(
+      cxxMemberCallExpr(
+          callee(memberExpr(hasDeclaration(cxxMethodDecl(hasName("size")))))),
+      callExpr(callee(
+          functionDecl(hasAnyName("::std::size", "::std::ranges::size")))));
```
Both of these are not constrained to be the span object. You'll probably want to swap the `callee` and `on` matchers, and add an `equalsBoundNode("span_object")` check for the `on` and `hasArgument(0,...` in the matchers of `SizeCall`.

Right now, this would match `span.subspan(other.size() - n)` and turn it into `span.last(n)`.

hjanuschka replied:
i'v been banging my head against this problem the whole weekend, with no success 🙈
do you have any existing check in mind that has a similar use case `span.subspan(other.size() - n)`

5chmidti replied:
This has generally been done with `utils::areStatementsIdentical` or its matcher equivalent, which is not 

See https://github.com/llvm/llvm-project/blob/3c464d23682b0f9e6f70965e8f8f3861c9ba5417/clang-tools-extra/clang-tidy/modernize/UseStartsEndsWithCheck.cpp#L73-L75

https://github.com/llvm/llvm-project/blob/3c464d23682b0f9e6f70965e8f8f3861c9ba5417/clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp#L151-L152

https://github.com/llvm/llvm-project/blob/3c464d23682b0f9e6f70965e8f8f3861c9ba5417/clang-tools-extra/clang-tidy/utils/Matchers.h#L164-L168

https://github.com/llvm/llvm-project/blob/3c464d23682b0f9e6f70965e8f8f3861c9ba5417/clang-tools-extra/clang-tidy/utils/ASTUtils.h#L40-L41

You can add `isStatementIdenticalToBoundNode` to your `SizeCall` matcher, checking the argument or object to be equivalent to `span_object`, e.g., `on(expr(isStatementIdenticalToBoundNode("span_object")))`

hjanuschka replied:
🙇 - that did the trick, awesome!

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+    diag(FirstCall->getBeginLoc(), "prefer 'span::first()' over 'subspan()'")
+        << FixItHint::CreateReplacement(FirstCall->getSourceRange(),
+                                        Replacement);
+  }
```
When you're inside the `FirstCall` case, you can't enter the `LastCall` case, so just return at the end of this if-branch.

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+          .bind("first_subspan"),
+      this);
+
+  // Match span.subspan(size() - n) or span.subspan(std::ranges::size(span) - n)
```
nit: `size()` -> `span.size()`

hjanuschka replied:
Done.

5chmidti reviewed (commented):
> wasting effort would not be the problem (it never is; learning is the goal) - right now i just dont understand the difference, and therefore dont see the benefit, as far as i see it would introduce more complicated code to still handle templates.

`TK_IgnoreUnlessSpelledInSource` changes how the AST is seen by excluding AST nodes that are implicit, e.g., implicit template instantiations or implicit casts. This can in some cases improve the readability of the check by simplifying the matcher, or improve performance because not every implicit AST node, such as instantiated templates, are traversed and checked. The effect of this varies a lot, depending on what a check is trying to achieve. Checkout the [Traverse Mode](https://clang.llvm.org/docs/LibASTMatchersReference.html#:~:text=Traverse%20Mode) section on top of the AST matcher reference for more details, and importantly, a lot of examples on the differences.

For this check, the benefit will be performance, and not readability. It will be possible to skip template instantiations, of which there may be many because `span` is *the* generic view on contiguous memory (joined by `mdspan`). Overall, the checking done by this check is not too costly, so there would not be a massive difference.

Godbolt with the AST view and clang-query are great to iteratively build up a matcher, checkout https://godbolt.org/z/GrfMTxeGa for matching uninstantiated templates with `TK_IgnoreUnlessSpelledInSource`

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,104 @@
+          on(expr(HasSpanType).bind("span_object")),
+          hasArgument(0, binaryOperator(hasOperatorName("-"), hasLHS(SizeCall),
+                                        hasRHS(expr().bind("count")))),
+          argumentCountIs(1))
```
The `argumentCountIs` matcher should be the first this that is checked inside the `cxxMemberCallExpr` matcher, because that is a simple integer comparison that can exclude a lot of AST nodes from further matching. The same applies to the other `cxxMemberCallExpr` above.

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,104 @@
+        CharSourceRange::getTokenRange(Count->getSourceRange()),
+        *Result.SourceManager, Result.Context->getLangOpts());
+
+    std::string Replacement = SpanText.str() + ".last(" + CountText.str() + ")";
```
Please use `llvm::Twine`, as in line 80

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,104 @@
+      CharSourceRange::getTokenRange(SpanObj->getSourceRange()),
+      *Result.SourceManager, Result.Context->getLangOpts());
+
+  if (const auto *FirstCall =
```
Except for which of the two functions should replace the matched code, these two `if` statements are identical. You could bind both cases to the same name, retrieve the `CXXMemberCallExpr`, and compare the number of arguments to determine which case it is:

```c++
    const auto IsFirst = SubSpan->getNumArgs() == 2;
    const StringRef FirstOrLast = IsFirst ? "first" : "last";

    std::string Replacement =
        (SpanText + "." + FirstOrLast + "(" + CountText + ")").str();

    diag(SubSpan->getBeginLoc(), "prefer 'span::%0()' over 'subspan()'")
        << FirstOrLast
        << FixItHint::CreateReplacement(SubSpan->getSourceRange(), Replacement);
```

hjanuschka replied:
thank you - haven't thought of this, great idea and super improvement!

hjanuschka replied:
Done.

hjanuschka reviewed (commented):
ready for re-review

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,97 @@
+  if (!SpanObj)
+    return;
+
+  StringRef SpanText = Lexer::getSourceText(
+      CharSourceRange::getTokenRange(SpanObj->getSourceRange()),
+      *Result.SourceManager, Result.Context->getLangOpts());
```
Please move this variable down, just before it is used, somewhere around `CountText`.

5chmidti commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,97 @@
+#include "../utils/ASTUtils.h"
+#include "../utils/Matchers.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/RecursiveASTVisitor.h"
```
Please remove this unused include

5chmidti reviewed (commented):
You could try to implement the check with `TK_TraverseUnlessSpelledInSource` if you'd like, which would only add two more matchers via `addMatcher`, and the traversal kind function (https://godbolt.org/z/GrfMTxeGa). IMO, this can be done in a follow-up by someone else as well.

`TK_TraversUnlessSpelledInSource` would not only have performance implications, but also niche correctness improvements: 
```diff
struct IntSpan {
    IntSpan(int*);
    IntSpan subspan(int, int);
};

template <typename T>
void test_instantiations(T val) {
-    val.subspan(0, 5);
+    val.first(5); // `IntSpan` does not have `first`
}

void test_intantiations() {
    int array[] = {0, 1, 2, 3, 4};
    auto sp = std::span<int>{array, 5};
    test_instantiations(sp);
    IntSpan intsp = array;
    test_instantiations(intsp);
}
```

hjanuschka commented:
feedback addressed, you were right about the template case, it is really broken, adding:
```
  std::optional<TraversalKind> getCheckTraversalKind() const override {
    return TK_IgnoreUnlessSpelledInSource;
  }
  ```
  
fixes it, but now templates seem to be skipped completely. and existing tests are not "passing" anymore:

```
template <typename T>
void testTemplate() {
  T arr[] = {1, 2, 3, 4, 5};
  std::span<T> s(arr, 5);

  auto sub1 = s.subspan(0, 3);
  // CHECK-MESSAGES: :[[@LINE-1]]:15: warning: prefer 'span::first()' over 'subspan()'
  // CHECK-FIXES: auto sub1 = s.first(3);

  auto sub2 = s.subspan(s.size() - 2);
  // CHECK-MESSAGES: :[[@LINE-1]]:15: warning: prefer 'span::last()' over 'subspan()'
  // CHECK-FIXES: auto sub2 = s.last(2);

  __SIZE_TYPE__ n = 2;
  auto sub3 = s.subspan(0, n);
  // CHECK-MESSAGES: :[[@LINE-1]]:15: warning: prefer 'span::first()' over 'subspan()'
  // CHECK-FIXES: auto sub3 = s.first(n);

  auto sub4 = s.subspan(1, 2);  // No warning
  auto sub5 = s.subspan(2);     // No warning

  auto complex = s.subspan(0 + (s.size() - 2), 3);  // No warning

  auto complex2 = s.subspan(100 + (s.size() - 2));  // No warning
}
```

do you have any ideas?

5chmidti commented:
> but now templates seem to be skipped completely

Template instantiations are skipped, not their definition. The matchers no longer work, because the types you are working with are no longer necessarily concrete types, but instead they can be dependent. And so can be expressions. 

In addition to the matchers for non-dependent code, you need to add matchers for the dependent case. See godbolt link I had linked: https://godbolt.org/z/GrfMTxeGa
and the included matcher (example, not full matcher):

```
callExpr(                            // no longer a CXXMemberCallExpr
    callee(
        cxxDependentScopeMemberExpr( // the callee is now a dependent member
            hasMemberName("subspan"),
            hasObjectExpression(
                hasType(             // this part is effectively HasSpanType but for a dependent type of `std::span<T>`
                    elaboratedType(
                        namesType(
                            qualType(
                                hasCanonicalType(
                                    qualType(
                                        hasDeclaration(
                                            namedDecl(hasName("::std::span"))
                                        )
                                    )
                                )
                            )
                        )
                    )
                )
            )
        )
    )
)
```

hjanuschka committed:
[clang-tidy] Add readability-use-span-first-last check

Add new clang-tidy check that suggests using std::span's more expressive
first() and last() member functions instead of equivalent subspan() calls.

These dedicated methods were added to C++20 to provide clearer alternatives
to common subspan operations. They improve readability by better expressing
intent and are less error-prone by eliminating manual offset calculations.

For example:
  s.subspan(0, n)         -> s.first(n)
  s.subspan(s.size() - n) -> s.last(n)

Non-zero offset with count (like subspan(1, n)) or offset-only calls
(like subspan(n)) are not transformed as they have no clearer equivalent.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+(like ``subspan(n)``) have no clearer equivalent using ``first()`` or
+``last()``, so these cases are not transformed.
+
+This check is only active when C++20 or later is used.
```
```suggestion
This check is enabled for C++20 or later.
```

hjanuschka commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,110 @@
+//===--- UseSpanFirstLastCheck.cpp - clang-tidy-----------------*- C++ -*-===//
```
Done.

hjanuschka commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -145,6 +145,10 @@ New checks
 New check aliases
 ^^^^^^^^^^^^^^^^^
 
+- New check `readability-use-span-first-last` has been added that suggests using
```
Done.

hjanuschka commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,110 @@
+        Context.getSourceManager(), Context.getLangOpts());
+    const auto *Base =
+        cast<CXXMemberCallExpr>(Call)->getImplicitObjectArgument();
+    auto BaseStr = Lexer::getSourceText(
```
Done.

hjanuschka commented on clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst:
```diff
@@ -0,0 +1,24 @@
+
+Covered scenarios:
+
+=============================== ================
```
Done.

hjanuschka commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.h:
```diff
@@ -0,0 +1,43 @@
+///   auto sub4 = s.subspan(n);              // not changed
+/// \endcode
+///
+/// The check is only active in C++20 mode.
```
Done.

zeyi2 commented on clang-tools-extra/clang-tidy/readability/UseSpanFirstLastCheck.cpp:
```diff
@@ -0,0 +1,121 @@
+          .bind("subspan_call"),
+      this);
+
+  // Match span.subspan(span.size() - n) -> last(n)
```
Can we also support this pattern?
```
s.subspan(s.size() - n, n) -> last(n)
```
